### PR TITLE
TEST-#3422: always import read data to OmniSci in IO tests

### LINKED ISSUE #3422: OmniSci backend IO tests don't actually involve OmniSci
In the current IO testing flow, the read Arrow table is not actually being imported in the OmniSci. We need to explicitly import read data to OmniSci to verify that everything is working properly.

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

This PR adds explicit import of the read data to OmniSci at every IO test. Modification of the `eval_io` function was used where possible, in other places it was done by the `ForceOmnisciImport` context class. Both of these were placed at the newly created utility file `experimental/engines/omnisci_on_native/test/utils.py`, other testing helper functions that previously were located at the file with tests were also moved to the `utils.py`.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #3422 
- [x] tests <s>added and</s> passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/pandas/test/utils.py [hunk 1]
```diff
@@ -773,6 +773,7 @@ def call_eval_general():
             pd,
             pandas,
             applyier,
+            comparator=comparator,
```
Comment (dchigarev) [inline modin/pandas/test/utils.py:776]: it appeared that this parameter hadn't been forwarding

File: modin/experimental/engines/omnisci_on_ray/test/utils.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        raising_exceptions=raising_exceptions,
+        check_kwargs_callable=check_kwargs_callable,
+        modin_warning=modin_warning,
+        md_extra_kwargs=md_extra_kwargs,
+        *args,
+        **kwargs,
+    )
+
+
+def trigger_import(*dfs):
```
Comment (dchigarev) [inline modin/experimental/engines/omnisci_on_ray/test/utils.py:68]: this is almost a full copy of the [`asv_bench.benchmarks.utils.common.trigger_import`](https://github.com/modin-project/modin/blob/0a642753af2cd2332ac71166bcddc89b0cace9d5/asv_bench/benchmarks/utils/common.py#L443-L468) function. We can't import that function from `asv_bench`, since ASV stuff is not a part of the modin package, we can't import this function to `asv_bench` from here either, because this will break ASV compatibility with the previous version of modin.

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py [hunk 3]
```diff
@@ -186,11 +109,13 @@ def test_housing_csv(self):
                 "dtype": self.boston_housing_dtypes,
             },
         ):
-            rp = pandas.read_csv(csv_file, **kwargs)
-            rm = to_pandas(pd.read_csv(csv_file, engine="arrow", **kwargs))
-            assert rp is not None
-            assert rm is not None
-            # TODO: df_equals(rp, rm)  #  needs inexact comparison
```
Comment (dchigarev) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:193]: inexact comparison appeared to be unneeded, output frames are identical

File: modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py [hunk 4]
```diff
@@ -251,6 +188,7 @@ def test_null_col(self, null_dtype):
             dtype={"a": "int64", "b": "int64", "c": null_dtype},
             skiprows=1,
         )
+        trigger_import(exp)
```
Comment (ienkovich) [inline modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:191]: Due to the following computations, we will have execution and therefore import triggered anyway for `exp`.

Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py]: moved import triggering to the place before casting to pandas and also added a check to do not import the same frame multiple times

**Comment by dchigarev** [comment]
@modin-project/modin-omnisci @modin-project/modin-core just a reminder to review

**Comment by amyskov** [comment]
Looks great! A little bit confusing that imported to OmniSci table doesn't used during pandas/Modin results comparison by the `comparator`, shouldn't it be exported back for checking?

**Comment by ienkovich** [comment]
> Looks great! A little bit confusing that imported to OmniSci table doesn't used during pandas/Modin results comparison by the `comparator`, shouldn't it be exported back for checking?

We keep arrow tables for imported partitions and therefore never need to export them back.

**Comment by dchigarev** [comment]
After adding modified `eval_io` to the `pandas/test/test_io.py` with ~500 `read_csv` tests it appeared that the `test_io.py::TestCsv` step is failing with an out-of-memory error. I've added a print statement that prints the amount of used system RAM in tests, the job is being killed shortly after the RAM usage [hits 100%](https://github.com/modin-project/modin/pull/3423/checks?check_run_id=3633094484#step:8:1384).

The reason for hitting 100% RAM usage is that this step has a lot of `read_csv` calls, which generates a lot of tables in OmniSci, which are never going to be deleted because there is no mechanism of garbage collecting of orphaned tables.

@ienkovich is this awared problem of no garbage collection of unused tables in OmniSci and is there a way to solve this problem in the context of this PR (something like cycle-rewriting or similar)?

**Comment by dchigarev** [comment]
In order to resolve RAM overflow by Omnisci tables I've added a mechanism of dropping them at the end of every test that requires forcibly import read data to Omnisci (done by [`ForceOmnisciImport`](https://github.com/dchigarev/modin/blob/40fd7d71f57341bdabf5120ee77db5431c32cdb6/modin/experimental/engines/omnisci_on_native/test/utils.py#L69) context-class). 

After that, it revealed that one of the IO tests consumes ~17Gb of RAM when data is imported to Omnisci (mentioned test is [`test_read_csv_default_to_pandas_url`](https://github.com/dchigarev/modin/blob/40fd7d71f57341bdabf5120ee77db5431c32cdb6/modin/pandas/test/test_io.py#L971)). The reason for that is that in this test it reads a many-column frame (~1000 columns) where each column contain only strings. It seems that it's really unoptimized case for Omnisci, so I've added skipping this test if available RAM is less than 17Gb and selected backend is Omnisci.

**Comment by dchigarev** [comment]
@amyskov @ienkovich @modin-project/modin-core PR is ready for review

File: modin/pandas/test/test_io.py [hunk 5]
```diff
@@ -952,6 +957,13 @@ def test_read_csv_default_to_pandas(self):
                 StringIO(open(pytest.csvs_names["test_read_csv_regular"], "r").read())
             )
 
+    @pytest.mark.skipif(
+        condition=(
+            Backend.get() == "Omnisci"
+            and psutil.virtual_memory().available < 17 * 2 ** 30
+        ),
+        reason="It takes about ~17Gb of RAM for Omnisci to import table from this test because of too many (~1000) string columns in it.",
```
Comment (anmyachev) [inline modin/pandas/test/test_io.py:965]: We can use subset of all columns.

Comment (dchigarev) [reply modin/pandas/test/test_io.py]: done

**Review by anmyachev — DISMISSED** [review]

File: modin/experimental/engines/omnisci_on_native/test/utils.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+                frame_id = OmnisciServer().put_arrow_to_omnisci(frame)
+            else:
+                raise TypeError(
+                    f"Unexpected storage format, expected pandas.DataFrame or pyarrow.Table, got: {type(frame)}."
+                )
+            partition.frame_id = frame_id
+            self._imported_frames.append((df, frame_id))
+
+    def __enter__(self, *dfs):
+        pass
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_native/test/utils.py:105]: it should be returning `self` so things like
```python
with ForceOmnisciImport(df) as foo:
    print(foo)
```
work correctly. Not that it is required currently, but it's better to implement a general context manager protocol from the get go.

File: modin/experimental/engines/omnisci_on_native/test/utils.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        for df, frame_id in self._imported_frames:
+            actual_frame_id = df._query_compiler._modin_frame._partitions[0][0].frame_id
+            OmnisciServer().executeDDL(f"DROP TABLE IF EXISTS {frame_id}")
+            if actual_frame_id == frame_id:
+                df._query_compiler._modin_frame._partitions[0][0].frame_id = None
+        self._imported_frames = []
+
+
+def set_execution_mode(frame, mode, recursive=False):
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_native/test/utils.py:116]: can we have a small docstring here, at least describing what this does and why it's needed?

File: modin/experimental/engines/omnisci_on_native/test/utils.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+                fn=fn,
+                data=data,
+                data2=data2,
+                force_lazy=force_lazy,
+                force_arrow_execute=force_arrow_execute,
+                allow_subqueries=allow_subqueries,
+                constructor_kwargs=constructor_kwargs,
+                **kwargs,
+            )
+            _ = exp_res.index
```
Comment (vnlitvinov) [inline modin/experimental/engines/omnisci_on_native/test/utils.py:183]: why is this needed?

Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_native/test/utils.py:183]: it seems that it was done in order to trigger actual computations (to get the index axis we need the frame to be computed)

BTW, I haven't changed this function as well as `set_execution_mode`, these were just copied here from the `test_dataframe.py`

File: modin/experimental/engines/omnisci_on_native/test/utils.py (hunk 7 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_native/test/utils.py:116]: added

File: modin/experimental/engines/omnisci_on_native/test/utils.py (hunk 6 above)
Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_native/test/utils.py]: done

**Review by ienkovich — APPROVED** [review]

**Review by vnlitvinov — APPROVED** [review]
LGTM, thanks!

File: modin/experimental/engines/omnisci_on_native/test/utils.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    def __init__(self, *dfs):
+        self._imported_frames = []
+        for df in dfs:
+            if not isinstance(df, (pd.DataFrame, pd.Series)):
+                continue
+            df.shape  # to trigger real execution
+            if df.empty:
+                continue
+            partition = df._query_compiler._modin_frame._partitions[0][0]
+            if partition.frame_id is not None:
```
Comment (YarShev) [inline modin/experimental/engines/omnisci_on_native/test/utils.py:90]: Shouldn't this be like `if partition.frame_id is None:`?

Comment (dchigarev) [reply modin/experimental/engines/omnisci_on_native/test/utils.py:90]: `frame_id` holds the table name that is imported to OmniSci, which corresponds to the `partition` DataFrame. If `frame_id` is None, then no import has been triggered yet, so we have to stay in this for loop and import the table, if `frame_id` is not None, then the table has been already imported and we have to skip this frame.

Comment (YarShev) [reply modin/experimental/engines/omnisci_on_native/test/utils.py:90]: I got, thanks!
